### PR TITLE
Optional chaining

### PR DESCRIPTION
**Pull request by mariusGundersen**
This adds support for the three types of optional chaining:

```js
a?.b // returns b unless a is nullish
a?.() // calls the function a unless a is nullish
a?.[0] // returns the first item in the array a unless a is nullish
```

### CODE REVIEW

**Comment by mariusGundersen** [comment]
This is wished for here: https://github.com/espruino/Espruino/issues/1302#issuecomment-655780267

File: src/jsparse.c [hunk 1]
```diff
@@ -783,10 +783,10 @@ NO_INLINE JsVar *jspeFunctionCall(JsVar *function, JsVar *functionName, JsVar *t
...
 
             if (execInfo.execute&EXEC_DEBUGGER_STEP_INTO) {
-              if (functionName)
-                jsiConsolePrintf("Stepping into %v\n", functionName);
-              else
-                jsiConsolePrintf("Stepping into function\n", functionName);
+	      if (functionName)
+		jsiConsolePrintf("Stepping into %v\n", functionName);
+	      else
+		jsiConsolePrintf("Stepping into function\n", functionName);
```
Comment (gfwilliams) [inline src/jsparse.c:789]: I just noticed these spaces got replaced with tabs. Please can you try and tweak these (and change your editor so tabs don't get added?). It really screws up the layout if someone views the code and doesn't have 8 space tabs set

File: src/jsparse.c [hunk 2]
```diff
@@ -1123,6 +1150,10 @@ NO_INLINE JsVar *jspeFactorMember(JsVar *a, JsVar **parentResult) {
         jsvUnLock(aVar);
       }
       jsvUnLock(index);
+
+#ifndef ESPR_NO_OPTIONAL_CHAINING
+      JSP_RESTORE_EXECUTE();
```
Comment (gfwilliams) [inline src/jsparse.c:1155]: I'm afraid this breaks stuff.  `undefined[1]` should cause an exception (like 1143) but JSP_RESTORE_EXECUTE removes the error flag, so it just quietly completes. I can imagine there might be other code you could come up with (not just errors, but return/break/etc) too

**Comment by gfwilliams** [comment]
Sorry it's taken so long to look into this! But I am a bit concerned about the use of `JSP_SAVE_EXECUTE/RESTORE` - they need to be used with real care so you don't accidentally remove execution state. I think it might be possible to make it work without them though, as all you care about is executing or not (rather than any of the other state)

**Pull request closed without merging**